crates/uv-platform-tags/src/abi_tag.rs

```diff
@@ -0,0 +1,445 @@
+            } => {
+                write!(
+                    f,
+                    "pyston{py_major}{py_minor}-pyston_{impl_major}{impl_minor}"
+                )
+            }
+        }
+    }
+}
+
+impl FromStr for AbiTag {
+    type Err = ParseAbiTagError;
+
+    /// Parse an [`AbiTag`] from a string.
+    #[allow(clippy::cast_possible_truncation)]
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        /// Parse a Python version from a string (e.g., convert `39` into `(3, 9)`).
+        fn parse_python_version(
+            version_str: &str,
+            implementation: &'static str,
+            full_tag: &str,
+        ) -> Result<(u8, u8), ParseAbiTagError> {
+            let major = version_str
+                .chars()
+                .next()
+                .ok_or_else(|| ParseAbiTagError::MissingMajorVersion {
+                    implementation,
+                    tag: full_tag.to_string(),
+                })?
+                .to_digit(10)
+                .ok_or_else(|| ParseAbiTagError::InvalidMajorVersion {
+                    implementation,
+                    tag: full_tag.to_string(),
+                })? as u8;
+            let minor = version_str
+                .get(1..)
+                .ok_or_else(|| ParseAbiTagError::MissingMinorVersion {
+                    implementation,
+                    tag: full_tag.to_string(),
+                })?
+                .parse::<u8>()
+                .map_err(|_| ParseAbiTagError::InvalidMinorVersion {
+                    implementation,
+                    tag: full_tag.to_string(),
+                })?;
+            Ok((major, minor))
+        }
+
+        /// Parse an implementation version from a string (e.g., convert `37` into `(3, 7)`).
+        fn parse_impl_version(
+            version_str: &str,
+            implementation: &'static str,
+            full_tag: &str,
+        ) -> Result<(u8, u8), ParseAbiTagError> {
+            let major = version_str
+                .chars()
+                .next()
+                .ok_or_else(|| ParseAbiTagError::MissingImplMajorVersion {
+                    implementation,
+                    tag: full_tag.to_string(),
+                })?
+                .to_digit(10)
+                .ok_or_else(|| ParseAbiTagError::InvalidImplMajorVersion {
+                    implementation,
+                    tag: full_tag.to_string(),
+                })? as u8;
+            let minor = version_str
+                .get(1..)
+                .ok_or_else(|| ParseAbiTagError::MissingImplMinorVersion {
+                    implementation,
+                    tag: full_tag.to_string(),
+                })?
+                .parse::<u8>()
+                .map_err(|_| ParseAbiTagError::InvalidImplMinorVersion {
+                    implementation,
+                    tag: full_tag.to_string(),
+                })?;
+            Ok((major, minor))
+        }
```

Comment:
I think my comments from above apply here too.

I _personally_ am a fan of very granular error reporting like you have here. I've just been bitten so many times by overly generic and unhelpful error messages that I like this granular style much more. Like, it's more code, but I don't see this changing much or otherwise being hard to maintain personally.